core/trino-main/src/main/java/io/trino/operator/exchange/PartitioningExchanger.java

```diff
@@ -58,9 +59,12 @@ public void accept(Page page)
     {
         Page partitionPage = partitionedPagePreparer.apply(page);
         // assign each row to a partition. The assignments lists are all expected to cleared by the previous iterations
+        if (partitions.length < partitionPage.getPositionCount()) {
+            partitions = new int[partitionPage.getPositionCount()];
+        }
+        partitionFunction.getPartitions(partitionPage, 0, partitionPage.getPositionCount(), partitions);
         for (int position = 0; position < partitionPage.getPositionCount(); position++) {
-            int partition = partitionFunction.getPartition(partitionPage, position);
-            partitionAssignments[partition].add(position);
+            partitionAssignments[partitions[position]].add(position);
```

Comment:
There's probably an improvement to be gained by fusing this loop with `partitionAssignments.getPartitions`, but we can worry about that as a follow up.